heartbeat/autodiscover/builder/hints/monitors.go

```diff
@@ -134,28 +141,46 @@ func (hb *heartbeatHints) getProcessors(hints mapstr.M) []mapstr.M {
 	return utils.GetConfigs(hints, "", "processors")
 }
 
-func (hb *heartbeatHints) getHostsWithPort(hints mapstr.M, port int) []string {
-	var result []string
+func (hb *heartbeatHints) getHostsWithPort(hints mapstr.M, port int) ([]string, error) {
 	thosts := utils.GetHintAsList(hints, "", hosts)
-	// Only pick hosts that have ${data.port} or the port on current event. This will make
-	// sure that incorrect meta mapping doesn't happen
+	mType := utils.GetHintString(hints, "", scheme)
+
+	// We can't reliable detect duplicated monitors since we don't have all ports/hosts,
+	// relying on scheduler deduping monitors, see https://github.com/elastic/beats/pull/29041
+	hostSet := map[string]struct{}{}
 	for _, h := range thosts {
-		if strings.Contains(h, "data.port") || strings.Contains(h, fmt.Sprintf(":%d", port)) ||
-			// Use the event that has no port config if there is a ${data.host}:9090 like input
-			(port == 0 && strings.Contains(h, "data.host")) {
-			result = append(result, h)
-		} else if port == 0 && !strings.Contains(h, ":") {
-			// For ICMP like use cases allow only host to be passed if there is no port
-			result = append(result, h)
-		} else {
-			hb.logger.Warn("unable to frame a host from input host: %s", h)
+		if mType == "icmp" && strings.Contains(h, ":") {
+			hb.logger.Warnf("ICMP scheme does not support port specification: %s", h)
+			continue
+		} else if strings.Contains(h, ":") && strings.Contains(h, "data.port") && port == 0 {
```

Comment:
are we removing the check for `data.host` altogether?

Reply:
also I don't think we should check for `data.port` in any part of the string. We should check it for `:${data.port}`

Reply:
> are we removing the check for data.host altogether?

IMO, the complicated thing here is building a general case. Pods are usually addresed via ip, services can be addressed by dns, some monitors might make sense for a single pod, others will be broken for a replica set. 

@andrewvc I think we have two paths to solve this, wdyt?:
- allow any value combination in without trying to match host-port, should fix errors like the ones listed [here](https://github.com/elastic/beats/issues/17771) with the downside that it can have unexpected results to be handled by the user, eg: generating a unique monitor config for a replica set pods will delete the monitor if a pod is removed from the set.
- Fix the empty host error and effectively have an even stricter matching host-port approach.

Reply:
Can you provide some more detail on the stricter matching approach? What would that look like?

Reply:
Sure, it would involve making sure that the monitor generated is dynamic in relation to the hint event data. It's basically what we do now by checking `${data.host}` and `${data.port}` but fixing the empty config scenario we have.

The problem with this approach is that it prevents some common use cases, that's why there's so many request to remove the restriction altogether.